packages/core/src/utils/GroupGraph.ts

```diff
@@ -0,0 +1,103 @@
+import { Shape, ShapeAD } from "../types/shape";
+import Graph from "./Graph";
+
+export type GroupGraphNode<T extends Shape | ShapeAD> = {
+  shape: T;
+  index: number;
+};
+
+export type GroupGraph<T extends Shape | ShapeAD> = Graph<
+  string,
+  GroupGraphNode<T>
+>;
+
+const makeGroupGraphNode = <T extends Shape | ShapeAD>(
+  shape: T,
+  index: number
+): [string, GroupGraphNode<T>, string[]] => {
+  const shapeType = shape.shapeType;
+  const shapeNameVal = shape.properties["name"];
+  if (shapeNameVal.tag !== "StrV") {
+    throw Error("Shape name is not a string");
+  }
+  const shapeName = shapeNameVal.contents;
+
+  if (shapeType === "Group") {
+    const subGPIsVal = shape.properties["shapes"];
+    if (subGPIsVal.tag !== "ShapeListV") {
+      throw Error("Group content not a list of shapes");
+    }
+    const subGPIs = subGPIsVal.contents;
+    const subShapeNames = subGPIs.map((gpi) => {
+      const subShapeNameVal = gpi.contents[1]["name"];
+      if (subShapeNameVal.tag !== "StrV") {
+        throw Error("Shape name is not a string");
+      }
+      return subShapeNameVal.contents;
+    });
+    return [
+      shapeName,
+      {
+        shape,
+        index,
+      },
+      subShapeNames,
+    ];
+  } else {
+    return [
+      shapeName,
+      {
+        shape,
+        index,
+      },
+      [],
+    ];
+  }
+};
+
+export const makeGroupGraph = <T extends Shape | ShapeAD>(
+  shapes: T[]
+): GroupGraph<T> => {
+  const graph: GroupGraph<T> = new Graph();
+  const rawGraph: { [k: string]: [GroupGraphNode<T>, string[]] } = {};
+  for (let i = 0; i < shapes.length; i++) {
+    const [name, node, children] = makeGroupGraphNode(shapes[i], i);
+    rawGraph[name] = [node, children];
+  }
+
+  for (const [name, [node, children]] of Object.entries(rawGraph)) {
+    graph.setNode(name, node);
+    for (const child of children) {
+      graph.setNode(child, rawGraph[child][0]);
+      graph.setEdge({ i: name, j: child, e: undefined });
+    }
```

Comment:
Why do you need `rawGraph`? Can you insert nodes to the graph immediately after calling `makeGroupGraphNode`?

Reply:
I originally needed it, but not anymore. Done.